Fix Windows build with --enable-werror

hebasto opened the pull request:
This PR makes possible to cross-compile Windows build with `--enable-werror --enable-suppress-external-warnings`.
Some problems are fixed, others are silenced.

Also `--enable-werror` is enabled for Cirrus CI Windows build (the last one on Cirrus CI without `--enable-werror`).

hebasto commented:
cc @fanquake @MarcoFalke @vasild

vasild commented on src/fs.cpp:
```diff
@@ -148,7 +148,10 @@ std::string get_filesystem_error_message(const fs::filesystem_error& e)
 
 // reference: https://github.com/gcc-mirror/gcc/blob/gcc-7_3_0-release/libstdc%2B%2B-v3/include/std/fstream#L270
-
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wswitch"
+#endif
 static std::string openmodeToStr(std::ios_base::openmode mode)
```
This will disable `-Wswitch` for any GCC compiler (even latest on Linux). Shouldn't we restrict that somehow? Maybe check if `WIN32` is defined or for a specific GCC version?

hebasto replied:
https://github.com/bitcoin/bitcoin/blob/06ff47560ecb1d6a80e9fbd2f169f6ed8f31c464/src/fs.cpp#L147-L148

vasild replied:
Oh, it is already inside `#ifdef WIN32`, sorry for the noise.

vasild commented on configure.ac:
```diff
@@ -392,7 +392,13 @@ if test "x$enable_werror" = "xyes"; then
+  dnl -Wreturn-type is broken in GCC for MinGW-w64.
+  dnl https://sourceforge.net/p/mingw-w64/bugs/306/
+  AX_CHECK_COMPILE_FLAG([-Werror=return-type], [ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=return-type"], [], [$CXXFLAG_WERROR],
+                        [AC_LANG_SOURCE([[#include <cassert>
+                                          int f(){ assert(false); }]])])
```
This is good and will work fine.

I am just curious, does that mingw bug 306 manifest if `assert(false)` is replaced with `abort()`?

hebasto replied:
Did not dive in details, but it seems to be dealing with an internal attribute of noreturn.

laanwj commented:
I'm a bit divided on whether building on every platform/compiler combination 'with -Werror' is actually a goal of the project. I mean, I think this is impossible, and will result in a lot of warning fixing for the sake of warning fixing.

In general Werror is a curse.

MarcoFalke commented:
Tend to agree with @laanwj, especially if the workarounds will make already hard to understand files like fs.cpp even more verbose.

vasild reviewed (approved):
ACK 06ff4756

Sometimes important warnings are emitted only on some platforms, so I think it is good to use `-Werror` on as most platforms as possible.

If keeping Windows builds warning-free turns out to be too big burden then the `NO_WERROR=1` can be added back.

hebasto commented:
> I'm a bit divided on whether building on every platform/compiler combination 'with -Werror' is actually a goal of the project. I mean, I think this is impossible, and will result in a lot of warning fixing for the sake of warning fixing.
> 
> In general Werror is a curse.

> Tend to agree with @laanwj, especially if the workarounds will make already hard to understand files like fs.cpp even more verbose.

Fair enough.

The goal of this PR (combined with #20544) is to prevent changes that introduce new warnings.

Anyway, feel free to close this.

practicalswift commented:
Concept ACK

hebasto commented:
@MarcoFalke 
> Tend to agree with @laanwj, especially if the workarounds will make already hard to understand files like fs.cpp even more verbose.

Only one-line change in `fs.cpp` now.

vasild commented on src/fs.cpp:
```diff
@@ -147,11 +147,11 @@ std::string get_filesystem_error_message(const fs::filesystem_error& e)
 static std::string openmodeToStr(std::ios_base::openmode mode)
 {
-    switch (mode & ~std::ios_base::ate) {
+    switch (static_cast<int>(mode & ~std::ios_base::ate)) { // casting is required to suppress -Wswitch warnings
```
`std::ios_base::ate` and friends are defined as `unsigned int` on my platform. What is the warning that this tries to fix?

hebasto replied:
> `std::ios_base::ate` and friends are defined as `unsigned int` on my platform.

https://github.com/gcc-mirror/gcc/blob/releases/gcc-7.3.0/libstdc%2B%2B-v3/include/bits/ios_base.h#L111-L122

> What is the warning that this tries to fix?

https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Warning-Options.html#Warning-Options:
> `-Wswitch`
> ... `case` labels outside the enumeration range also provoke warnings when this option is used (even if there is a `default` label).

vasild replied:
Ok, we are fine because if some of the case values is out of the range of `int`, then we would get a warning:
```cpp
void f(unsigned int x)
{
    const unsigned int y = 3000000000U;

    switch (static_cast<int>(x)) { 
    case y:
        std::cout << "foo\n";
        break;
    } 
}
```

```
error: case value evaluates to 3000000000, which cannot be narrowed to type 'int' [-Wc++11-narrowing]
```

from clang 12 with `-Wall`

vasild reviewed (approved):
ACK 926f6d82a31b2d2c82b830b2b672fa38890b8590

luke-jr commented on src/fs.cpp:
```diff
@@ -147,11 +147,11 @@ std::string get_filesystem_error_message(const fs::filesystem_error& e)
 static std::string openmodeToStr(std::ios_base::openmode mode)
 {
-    switch (mode & ~std::ios_base::ate) {
+    switch (static_cast<int>(mode & ~std::ios_base::ate)) { // casting is required to suppress -Wswitch warnings
```
NACK, the standard doesn't guarantee this fits in an `int`...

Can you use a pragma to disable the warning here?

hebasto replied:
[Done](https://github.com/bitcoin/bitcoin/pull/20586#issuecomment-748499647).

hebasto commented:
Reverted back 06ff47560ecb1d6a80e9fbd2f169f6ed8f31c464 <- 926f6d82a31b2d2c82b830b2b672fa38890b8590 ([pr20586.01](https://github.com/hebasto/bitcoin/commits/pr20586.01) <- [pr20586.02](https://github.com/hebasto/bitcoin/commits/pr20586.02)) as was [requested](https://github.com/bitcoin/bitcoin/pull/20586#discussion_r542049306) by @luke-jr.

This variant has been already [ACKed](https://github.com/bitcoin/bitcoin/pull/20586#pullrequestreview-546132266) by @vasild.

practicalswift commented:
cr ACK 06ff47560ecb1d6a80e9fbd2f169f6ed8f31c464: patch looks correct

FWIW, having this in `master` would have prevented the issue #21355 :)

MarcoFalke commented on src/fs.cpp:
```diff
@@ -148,7 +148,10 @@ std::string get_filesystem_error_message(const fs::filesystem_error& e)
-
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wswitch"
```
Is this commit " Fix Windows build with --enable-werror on Ubuntu Bionic (gcc 7.5) " still needed?

We no longer use mingw on bionic (only focal)

hebasto replied:
Right. #21036 made it possible.

Going to rebase the pr and remove that commit.

hebasto replied:
@MarcoFalke It still required to prevent `-Werror=switch` warnings on Focal.

Therefore, technically, no changes are required in the patch.

hebasto committed:
Fix Windows build with --enable-werror on Ubuntu Focal

hebasto committed:
ci: Make Cirrus CI Windows build with --enable-werror

hebasto commented:
Updated 06ff47560ecb1d6a80e9fbd2f169f6ed8f31c464 -> b367745cfe19f6de3f44b3adc90fa08e36e44bb6 ([pr20586.01](https://github.com/hebasto/bitcoin/commits/pr20586.01) -> [pr20586.03](https://github.com/hebasto/bitcoin/commits/pr20586.03)):
- rebased due to the conflict with #21467 
- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20586#discussion_r588848569)

vasild reviewed (approved):
ACK b367745cfe19f6de3f44b3adc90fa08e36e44bb6

jarolrod reviewed (commented):
ACK b367745cfe19f6de3f44b3adc90fa08e36e44bb6

Tested on Linux cross-compiling to Windows passing `--enable-werror` to configure

**Master:**
```
script/standard.cpp:215:26: error: control reaches end of non-void function [-Werror=return-type]
  215 |     std::vector<valtype> vSolutions;
```

**PR:**
Builds fine and built binaries work fine on `Windows 10`

practicalswift commented:
cr ACK b367745cfe19f6de3f44b3adc90fa08e36e44bb6: patch looks correct

hebasto commented:
@laanwj @MarcoFalke @luke-jr 

Do you mind having another look at this PR?

MarcoFalke commented on src/fs.cpp:
```diff
@@ -154,7 +154,10 @@ std::string get_filesystem_error_message(const fs::filesystem_error& e)
-
+#if defined(__GNUC__) && !defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wswitch"
```
Will this be needed after the boost::fs removal?

hebasto replied:
I think no, as this bunch of code should go: https://github.com/bitcoin/bitcoin/blob/d6e0d78c31557660274ef53cac912c468eecbe2d/src/fs.h#L66-L69
